google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java

```diff
@@ -191,15 +191,14 @@ protected void doStart() {
 
   @Override
   protected void doStop() {
-    runShutdown();
-
     lock.lock();
     try {
       clientStream.closeSendWithError(Status.CANCELLED.asException());
     } finally {
       lock.unlock();
       notifyStopped();
     }
+    runShutdown();
```

Comment:
We don't want to notifyStopped() until runShutdown() has completed because this has a lot of blocking work.

Reply:
Done. This is the order now:
1. Close stream
2. Wait for message dispatcher to complete (dispatch message to end user, and wait for ack from end user)
3. Wait for ack operations to complete (send ack to pubsub)
4. notifyStop